test/fuzz/transformation_changing_memory_semantics_test.cpp

```diff
@@ -0,0 +1,482 @@
+               OpReturn
+               OpFunctionEnd
+  )";
+  const auto env = SPV_ENV_UNIVERSAL_1_3;
+  const auto consumer = nullptr;
+  const auto context = BuildModule(env, consumer, shader, kFuzzAssembleOption);
+  spvtools::ValidatorOptions validator_options;
+  ASSERT_TRUE(fuzzerutil::IsValidAndWellFormed(context.get(), validator_options,
+                                               kConsoleMessageConsumer));
+  TransformationContext transformation_context(
+      MakeUnique<FactManager>(context.get()), validator_options);
+
+  // Bad: Instruction does not exist.
+  ASSERT_FALSE(TransformationChangingMemorySemantics(
+                   MakeInstructionDescriptor(150, SpvOpAtomicLoad, 0), 0, 100)
+                   .IsApplicable(context.get(), transformation_context));
+
+#ifndef NDEBUG
+
+  // Bad: Operand index does not equal 0 or 1.
+  ASSERT_DEATH(TransformationChangingMemorySemantics(
+                   MakeInstructionDescriptor(21, SpvOpAtomicLoad, 0), 2, 100)
+                   .IsApplicable(context.get(), transformation_context),
+               "The instruction may not be an atomic or barrier instruction. \
+                The operands index may be not equal 0 or 1. \
+                The index may be equal to one and the expected is zero.");
+
+  // Bad: The SpvOpAtomicLoad takes one memory semantic value and the operand
+  // index passed is 1.
+  ASSERT_DEATH(TransformationChangingMemorySemantics(
+                   MakeInstructionDescriptor(21, SpvOpAtomicLoad, 0), 1, 100)
+                   .IsApplicable(context.get(), transformation_context),
+               "The instruction may not be an atomic or barrier instruction. \
+                The operands index may be not equal 0 or 1. \
+                The index may be equal to one and the expected is zero.");
+#endif
+  // Bad: The new Id instruction does not exist.
+  ASSERT_FALSE(TransformationChangingMemorySemantics(
+                   MakeInstructionDescriptor(21, SpvOpAtomicLoad, 0), 0, 101)
+                   .IsApplicable(context.get(), transformation_context));
+
+  // Bad: The new Id instruction exists, but not OpConstant.
+  ASSERT_FALSE(TransformationChangingMemorySemantics(
+                   MakeInstructionDescriptor(21, SpvOpAtomicLoad, 0), 0, 13)
+                   .IsApplicable(context.get(), transformation_context));
+
+  // Bad: The OpAtomicStore memory semantics value cannot be AcquireRelease.
+  ASSERT_FALSE(TransformationChangingMemorySemantics(
+                   MakeInstructionDescriptor(24, SpvOpAtomicStore, 0), 0, 100)
+                   .IsApplicable(context.get(), transformation_context));
+
+  // Bad: The OpAtomicExchange(read-modify-write) memory semantics value cannot
+  // be changed from Acquire to Release.
+  ASSERT_FALSE(TransformationChangingMemorySemantics(
+                   MakeInstructionDescriptor(23, SpvOpAtomicExchange, 0), 0, 99)
+                   .IsApplicable(context.get(), transformation_context));
```

Comment:
Actually, according to our shared doc, we only wanted to support:

* RMW: None -> AcquireRelease -> SequentiallyConsistent

So, although the check passes, I think the comment is incorrect. IsApplicable returns false because the old value is not one of the RMW values we expect. And so, the check is not actually very interesting. You should probably change the OpAtomicExchange in the original SPIR-V module use a memory semantics value that we expect.

Reply:
Changed to:
```
  // Bad transformation:
  // OpAtomicExchange AcquireRelease | UniformMemory
  // to:
  // OpAtomicExchange None           | UniformMemory
```